[salty-flower] iP

salty-flower opened the pull request:
# Tach
> "The iP is based on the generic project called [Project Duke](https://nus-cs2103-ay2223s2.github.io/website/se-book-adapted/projectDuke/index.html), the overview of which is given below." ([source](https://nus-cs2103-ay2223s2.github.io/website/schedule/week2/project.html#1-learn-about-the-project))

Tach is a dumb command-line interface program and will have an equally dumb GUI very soon. It's,
- dumb at UI
- dumb at features
- dumb to learn
- dumb to use
- worked on by a [dumb guy](https://github.com/salty-flower)

All you need to do is,

1. view the PR from [here](https://github.com/nus-cs2103-AY2223S2/ip/pull/213).
2. *click* on it.
3. and ~~tease~~ _interact with_ the author:laughing:.

And it is FREE!

Features:
- [X] accepting input of particular formats
- [X] generating junk files
- [X] giving dumb feedback
- [ ] managing tasks (will never do)
---
If you are a Java programmer, you can use it to practice Java too. Where is the `main` method? Not this one:
```java
public class Mian {
    public static void mian(String[] args) {
        Appliçation.launçh(MianApp.çlaşş, argş);
    }
}
```

Commit:
Add Gradle support

salty-flower committed:
:truck:Duke -> Tach

salty-flower committed:
:sparkles:New commands - Bye, Echo
:art:Primary skeleton

salty-flower committed:
:sparkles:New commands - AddTask, ShowTaskList
:sparkles:New service - TaskStorage
:truck:Command -> handlers
:truck:Introduce prefix "H" to handlers
:boom:Store handler object instead of class in Parser

salty-flower committed:
:sparkles:New commands - MarkTask

salty-flower committed:
:building_construction:Make classes final
:art:Auto-format

salty-flower committed:
:building_construction:Make Task abstract
:building_construction:Add 3 subtypes of Task
:sparkles:Add desired type of Task

salty-flower committed:
:white_check_mark:Add basic Text-UI diff test
:building_construction:Deprecate reflection

salty-flower committed:
:sparkles:Print error message for 2 cases
:building_construction:Support error handler in Parser
:building_construction:Add exception class
:recycle:Extract check-and-run method to utilities

salty-flower committed:
:sparkles:Support task deletion

salty-flower committed:
:recycle:Use ENUM for Task status (done/not yet)
:bulb:No ENUM for Task types; need Scala Case Classes

salty-flower committed:
:art:Format code
:art:Deprecate unused-warning suppression

salty-flower committed:
:pencil2:Unify regex modifiers to `private static final`

salty-flower committed:
:art::truck:
:art:Move data classes into `types.data`
:art:Separate `utilities` package from `services` for static helper classes
:truck:`Parser`->`Dispatcher`
:truck:`H***`->`J***` - _handler_ -> _job_
:truck:`Command`->`Handler`

salty-flower committed:
:sparkles:Save to file "./data/tasks.txt"

salty-flower committed:
:sparkles:Date as LocalDateTime

salty-flower committed:
Merge branch 'branch-Level-7'

salty-flower committed:
Merge branch 'branch-Level-8'

salty-flower committed:
Merge remote-tracking branch 'origin/add-gradle-support'

salty-flower committed:
:building_construction:
:building_construction:Separate interfaces for output and persistence
:building_construction:Extract StdOut, FileContainer as implementations

salty-flower committed:
:white_check_mark:Add 3 testcases

salty-flower committed:
:memo:Javadoc for half of the classes; left with containers and handlers.

salty-flower committed:
:art:Coding style improvements

salty-flower committed:
:sparkles:Add Find command

salty-flower committed:
Merge branch 'branch-A-JavaDoc'

salty-flower committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/services/Dispatcher.java

salty-flower committed:
Merge branch 'branch-Level-9'

thennant commented on src/main/java/handlers/JAddTask.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public final class JAddTask implements IHandler {
+    private static final Pattern TODO_PATTERN = Pattern.compile("(todo) (.*)");
+    private static final Pattern EVENT_PATTERN = Pattern.compile("(event) (.*) /from (.*) /to (.*)");
+    private static final Pattern DEADLINE_PATTERN = Pattern.compile("(deadline) (.*) /by (.*)");
+    private static final Pattern ALL_PATTERN = Pattern
+            .compile("(todo) (.*)|(deadline) (.*) /by (.*)|(event) (.*) /from (.*) /to (.*)");
+    private final TaskList ts;
+
```
I like your use of regex to set the formatting of the strings

thennant commented on src/test/java/handlers/JEchoTest.java:
```diff
@@ -0,0 +1,16 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class JEchoTest {
+    @Test
+    public void test1() {
+        String s = "what ever you input";
+        JEcho echo = new JEcho();
+        assertTrue(echo.canTake(s));
+        assertEquals(s, echo.take(s));
+    }
```
Sufficient test case provided

thennant commented on src/main/java/types/data/Event.java:
```diff
@@ -0,0 +1,40 @@
+     * @param d2 Ending date or time.
+     * @return Newly created Event object.
+     */
+    public static Event create(String n, String d1, String d2) {
+        return new Event(n, d1, d2);
+    }
```
Might want to consider better naming for parameters so as to follow the coding standard?

thennant commented on src/main/java/handlers/JDeleteTask.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class JDeleteTask implements IHandler {
+    private static final Pattern PATTERN = Pattern.compile("delete ([0-9]*)");
+    private final TaskList ts;
+
+    public JDeleteTask(TaskList ts) {
+        this.ts = ts;
+    }
+
+    @Override
+    public String take(String s) {
+        Matcher m = PATTERN.matcher(s);
+        if (!m.matches()) {
+            return "";
+        }
+        int no = Integer.parseInt(m.group(1));
+        if (no <= 0 || no > ts.getTaskCount()) {
+            return "Invalid task number.\n";
+        } else {
+            Task removed = ts.getTaskByNo(no);
+            ts.deleteByNo(no);
+            return String.format("Noted. I've removed this task:\n%s\nNow you have %d task(s) in the list.\n", removed,
+                    ts.getTaskCount());
+        }
+    }
```
Code looks neat and clean!

thennant reviewed (commented):
Hi there!
Overall, there was good use of OOP in your code. The classes were also well packaged. Your code also looks very clean and neat. Just some minor issues with how you name your parameters. Overall job well done!

salty-flower committed:
:art:Add checkstyle

salty-flower committed:
:sparkles:Add GUI (although lame)

salty-flower committed:
Merge branch 'branch-A-CheckStyle'

salty-flower committed:
Merge branch 'branch-Level-10'

# Conflicts:
#	src/main/java/Tach.java

salty-flower committed:
:art:Reorder import statement

hj-prk commented on src/main/java/handlers/JFind.java:
```diff
@@ -0,0 +1,48 @@
+            }
+        }
+
+        int j = 1;
```
Might be better to give it a different name?

hj-prk commented on src/main/java/handlers/JMarkTask.java:
```diff
@@ -0,0 +1,41 @@
+        this.ts = ts;
+    }
+
+    @Override
```
Needs comments what does it take

hj-prk commented on src/main/java/services/Dispatcher.java:
```diff
@@ -0,0 +1,100 @@
+
+        speakerRegistry.broadcast(defaultHandler.take(expr));
+    }
+
```
Good explanations on what methods do!

hj-prk reviewed (commented):
Neat codes! Awesome!

salty-flower committed:
:bento:Add assertions

Introduced assertions to multiple places across
the repo.

salty-flower committed:
:art:Improve code quality

Major concerns include
- make happy path prominent
- replace for-loops with streams when possible

Todo in later commits:
- extract methods for GUI
- introduce a string pool to avoid magic strings

salty-flower committed:
Merge pull request #1 from salty-flower/branch-A-CodeQuality

:art:Improve code quality

salty-flower committed:
Merge branch 'master' into branch-A-Assertions

salty-flower committed:
Merge pull request #2 from salty-flower/branch-A-Assertions

:bento:Add assertions

salty-flower committed:
:sparkles:Add snooze deadline

salty-flower committed:
:sparkles:Add snooze deadline

salty-flower committed:
Merge branch 'master' of https://github.com/salty-flower/ip

salty-flower committed:
:children_crossing:Improve GUI

Polish GUI by adding images with FXML.

salty-flower committed:
Merge commit '1bb132787f616d8ddfb2bd43879b9694db17c2b4'

salty-flower committed:
:children_crossing:Fix ellipsis

By specifying the minimum height of Label,
text could be displayed better.

salty-flower committed:
:memo:Add project screenshot

salty-flower committed:
:memo:Add brief user guide

salty-flower committed:
:children_crossing:Improve GUI

Auto adjust height of Label.

salty-flower committed:
:memo:Update screenshot

salty-flower committed:
:memo:Update documentation.

Added the `snooze` feature to user guide.
Updated several public method/class comments.
